Boost Error in ResourceLog

In combination with:
* GCC 7.3.1
* CUDA 9.2.88
* Boost 1.68.0 (C++11 build) OR Boost 1.65.1
* PIConGPU 0.4.0
* centos 7

I have checked the issue here:
https://github.com/ComputationalRadiationPhysics/picongpu/issues/2714#issue-367113446
But, there  is still such err in ResourceLog.hpp:
```
/trinity/home/b.lei/pic/src/blei/picongpu/include/pmacc/../picongpu/plugins/ResourceLog.hpp(115): here
/trinity/shared/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/stl_pair.h(331): error: identifier "std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_S_empty_rep_storage" is undefined in device code

/trinity/shared/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/stl_pair.h(331): error: identifier "std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_S_empty_rep_storage" is undefined in device code

2 errors detected in the compilation of "/tmp/tmpxft_0000926d_00000000-6_main.cpp1.ii".
CMake Error at picongpu_generated_main.cpp.o.Release.cmake:279 (message):
  Error generating file
  /trinity/home/b.lei/pic/build/CMakeFiles/picongpu.dir//./picongpu_generated_main.cpp.o
```

> @BifengLei in the meantime, you can skip the compilation of the `ResourceLog` plugin altogether by commenting out these two lines:
> 
> ```diff
> diff --git a/include/picongpu/plugins/PluginController.hpp b/include/picongpu/plugins/PluginController.hpp
> index ad6543d1c..e174c731f 100644
> --- a/include/picongpu/plugins/PluginController.hpp
> +++ b/include/picongpu/plugins/PluginController.hpp
> @@ -70,7 +70,7 @@
>  #endif
>  
>  #include "picongpu/plugins/Checkpoint.hpp"
> -#include "picongpu/plugins/ResourceLog.hpp"
> +// include "picongpu/plugins/ResourceLog.hpp"
>  
>  #include <pmacc/mappings/kernel/MappingDescription.hpp>
>  
> @@ -181,7 +181,7 @@ private:
>  #if (ENABLE_HDF5 == 1)
>          , plugins::multi::Master< hdf5::HDF5Writer >
>  #endif
> -        , ResourceLog
> +//        , ResourceLog
>      >;
> ```

@ax3l  Thank you so much for replying.
Unfortunately, there is still the err:
```
/trinity/home/b.lei/lib/gcc-7.3.1/boost-1.68.0/include/boost/regex/v4/basic_regex_parser.hpp(48): warning: integer conversion resulted in a change of sign

/trinity/shared/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/stl_pair.h(331): error: identifier "std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_S_empty_rep_storage" is undefined in device code

/trinity/shared/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/stl_pair.h(331): error: identifier "std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_S_empty_rep_storage" is undefined in device code

2 errors detected in the compilation of "/tmp/tmpxft_00018152_00000000-6_main.cpp1.ii".
CMake Error at picongpu_generated_main.cpp.o.Release.cmake:279 (message):
  Error generating file
  /trinity/home/b.lei/pic/build/CMakeFiles/picongpu.dir//./picongpu_generated_main.cpp.o
```

@ax3l Here is all of  build information:
```
-- Found CUDA: /trinity/shared/opt/cuda-9.2 (found suitable version "9.2", minimum required is "7.0") 
-- ALPAKA_ACC_GPU_CUDA_ONLY_MODE
-- ALPAKA_ACC_GPU_CUDA_ENABLED
-- _cupla_COMPILE_OPTIONS_PUBLIC: -std=c++11;-ftemplate-depth-512;-Wall;-Wextra;-Wno-unknown-pragmas;-Wno-unused-parameter;-Wno-unused-local-typedefs;-Wno-attributes;-Wno-reorder;-Wno-sign-compare
-- _cupla_COMPILE_DEFINITIONS_PUBLIC: ALPAKA_ACC_GPU_CUDA_ONLY_MODE;ALPAKA_ACC_GPU_CUDA_ENABLED;ALPAKA_DEBUG=0;CUPLA_STREAM_ASYNC_ENABLED=1
-- _cupla_INCLUDE_DIRECTORIES_PUBLIC: /trinity/shared/opt/cuda-9.2/include;/trinity/home/b.lei/pic/src/blei/picongpu/thirdParty/alpaka/include;/trinity/home/b.lei/pic/src/blei/picongpu/thirdParty/cupla/include
-- _cupla_LINK_LIBRARIES_PUBLIC: Boost::boost;general;/trinity/shared/opt/cuda-9.2/lib64/libcudart.so;general;pthread;general;rt
-- Boost version: 1.68.0
-- Found the following Boost libraries:
--   filesystem
--   system
--   math_tr1
-- Boost: result_of with TR1 style and decltype fallback
-- Found OpenMP_C: -fopenmp  
-- Found OpenMP_CXX: -fopenmp  
-- Found OpenMP: TRUE   
-- Found CUDA: /trinity/shared/opt/cuda-9.2 (found suitable version "9.2", minimum required is "5.0") 
-- Boost version: 1.68.0
-- Found OpenMP_C: -fopenmp  
-- Found OpenMP_CXX: -fopenmp  
-- Boost version: 1.68.0
-- Boost version: 1.68.0
-- Found the following Boost libraries:
--   program_options
--   regex
--   filesystem
--   system
--   math_tr1
--   serialization
-- Found CUDA: /trinity/shared/opt/cuda-9.2 (found version "9.2") 
-- Could NOT find NVML (missing: NVML_LIBRARY) 
-- Boost version: 1.68.0
-- Found the following Boost libraries:
--   program_options
-- Can NOT find 'adios_config' - set ADIOS_ROOT or check your PATH
-- Could NOT find ADIOS (missing: ADIOS_LIBRARIES ADIOS_INCLUDE_DIRS) (Required is at least version "1.13.1")
-- HDF5: Using hdf5 compiler wrapper to determine C configuration
-- Found Splash: /trinity/home/b.lei/lib/gcc-7.3.1/splash-1.68.0/lib/cmake/Splash
-- Found PNGwriter: /trinity/home/b.lei/lib/gcc-7.3.1/pngwriter/lib/cmake/PNGwriter
-- Could NOT find ISAAC - set ISAAC_DIR or check your CMAKE_PREFIX_PATH
-- Configuring done
-- Generating done
-- Build files have been written to: /trinity/home/b.lei/pic/build
[  4%] Built target cupla
[  8%] Linking CXX executable cuda_memtest
[ 20%] Built target mpiInfo
[ 50%] Built target picongpu-hostonly
/trinity/shared/opt/rh/devtoolset-7/root/usr/bin/../libexec/gcc/x86_64-redhat-linux/7/ld: warning: libmpi.so.12, needed by /trinity/home/b.lei/lib/gcc-7.3.1/zlib/lib/libz.so, may conflict with libmpi.so.40
[ 62%] Built target cuda_memtest
[ 66%] Building NVCC (Device) object CMakeFiles/picongpu.dir/picongpu_generated_main.cpp.o
/trinity/home/b.lei/lib/gcc-7.3.1/boost-1.68.0/include/boost/mpl/map/aux_/item.hpp(54): warning: "boost::mpl::aux::type_wrapper<T> operator/(const boost::mpl::m_item<Key, T, Base> &, boost::mpl::aux::type_wrapper<Key> *)" declares a non-template function -- add <> to refer to a template instance

/trinity/home/b.lei/lib/gcc-7.3.1/boost-1.68.0/include/boost/mpl/map/aux_/item.hpp(55): warning: "boost::mpl::aux::type_wrapper<boost::mpl::pair<Key, T>> operator|(const boost::mpl::m_item<Key, T, Base> &, boost::mpl::next<Base::order>::type *)" declares a non-template function -- add <> to refer to a template instance

/trinity/home/b.lei/lib/gcc-7.3.1/boost-1.68.0/include/boost/mpl/map/aux_/item.hpp(56): warning: "char (&operator||(const boost::mpl::m_item<Key, T, Base> &, boost::mpl::aux::type_wrapper<Key> *))[boost::mpl::next<Base::order>::type::value]" declares a non-template function -- add <> to refer to a template instance

/trinity/home/b.lei/lib/gcc-7.3.1/boost-1.68.0/include/boost/mpl/map/aux_/item.hpp(71): warning: "boost::mpl::aux::type_wrapper<mpl_::void_> operator/(const boost::mpl::m_mask<Key, Base> &, boost::mpl::aux::type_wrapper<Key> *)" declares a non-template function -- add <> to refer to a template instance

/trinity/home/b.lei/lib/gcc-7.3.1/boost-1.68.0/include/boost/mpl/map/aux_/item.hpp(72): warning: "boost::mpl::aux::type_wrapper<mpl_::void_> operator|(const boost::mpl::m_mask<Key, Base> &, boost::mpl::x_order_impl<Base, Key>::type *)" declares a non-template function -- add <> to refer to a template instance

/trinity/home/b.lei/pic/src/blei/picongpu/include/pmacc/../pmacc/particles/memory/dataTypes/Particle.hpp(167): warning: __device__ annotation is ignored on a function("operator=") that is explicitly defaulted on its first declaration

/trinity/home/b.lei/pic/src/blei/picongpu/include/pmacc/../pmacc/particles/memory/dataTypes/Particle.hpp(167): warning: __host__ annotation is ignored on a function("operator=") that is explicitly defaulted on its first declaration

/trinity/home/b.lei/pic/src/blei/picongpu/include/pmacc/../picongpu/fields/LaserPhysics.hpp(147): warning: pointless comparison of unsigned integer with zero

/trinity/home/b.lei/lib/gcc-7.3.1/boost-1.68.0/include/boost/fusion/support/unused.hpp(25): warning: __host__ annotation is ignored on a function("unused_type") that is explicitly defaulted on its first declaration

/trinity/home/b.lei/lib/gcc-7.3.1/boost-1.68.0/include/boost/fusion/support/unused.hpp(25): warning: __device__ annotation is ignored on a function("unused_type") that is explicitly defaulted on its first declaration

/trinity/home/b.lei/lib/gcc-7.3.1/boost-1.68.0/include/boost/fusion/support/unused.hpp(31): warning: __host__ annotation is ignored on a function("unused_type") that is explicitly defaulted on its first declaration

/trinity/home/b.lei/lib/gcc-7.3.1/boost-1.68.0/include/boost/fusion/support/unused.hpp(31): warning: __device__ annotation is ignored on a function("unused_type") that is explicitly defaulted on its first declaration

/trinity/home/b.lei/lib/gcc-7.3.1/boost-1.68.0/include/boost/fusion/container/vector/vector.hpp(189): warning: __host__ annotation is ignored on a function("vector_data") that is explicitly defaulted on its first declaration

/trinity/home/b.lei/lib/gcc-7.3.1/boost-1.68.0/include/boost/fusion/container/vector/vector.hpp(189): warning: __device__ annotation is ignored on a function("vector_data") that is explicitly defaulted on its first declaration

/trinity/home/b.lei/lib/gcc-7.3.1/boost-1.68.0/include/boost/fusion/container/vector/vector.hpp(276): warning: __host__ annotation is ignored on a function("vector") that is explicitly defaulted on its first declaration

/trinity/home/b.lei/lib/gcc-7.3.1/boost-1.68.0/include/boost/fusion/container/vector/vector.hpp(276): warning: __device__ annotation is ignored on a function("vector") that is explicitly defaulted on its first declaration

/trinity/home/b.lei/lib/gcc-7.3.1/boost-1.68.0/include/boost/regex/v4/basic_regex_parser.hpp(48): warning: integer conversion resulted in a change of sign

/trinity/shared/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/stl_pair.h(331): error: identifier "std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_S_empty_rep_storage" is undefined in device code

/trinity/shared/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/stl_pair.h(331): error: identifier "std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_S_empty_rep_storage" is undefined in device code

2 errors detected in the compilation of "/tmp/tmpxft_0005f9c5_00000000-6_main.cpp1.ii".
CMake Error at picongpu_generated_main.cpp.o.Release.cmake:279 (message):
  Error generating file
  /trinity/home/b.lei/pic/build/CMakeFiles/picongpu.dir//./picongpu_generated_main.cpp.o


make[2]: *** [CMakeFiles/picongpu.dir/picongpu_generated_main.cpp.o] Error 1
make[1]: *** [CMakeFiles/picongpu.dir/all] Error 2
make: *** [all] Error 2
```

--- a/INSTALL.rst
+++ b/INSTALL.rst
@@ -89,9 +89,9 @@ zlib
 
 boost
 """""
-- 1.62.0 - 1.68.0 (``program_options``, ``regex`` , ``filesystem``, ``system``, ``math``, ``serialization`` and header-only libs, optional: ``fiber`` with ``context``, ``thread``, ``chrono``, ``atomic``, ``date_time``)
+- 1.62.0 - 1.68.0 (``program_options``, ``filesystem``, ``system``, ``math``, ``serialization`` and header-only libs, optional: ``fiber`` with ``context``, ``thread``, ``chrono``, ``atomic``, ``date_time``,  ``regex``)
 - *note:* for CUDA 9+ support, use boost 1.65.1 or newer
-- *Debian/Ubuntu:* ``sudo apt-get install libboost-program-options-dev libboost-regex-dev libboost-filesystem-dev libboost-system-dev libboost-thread-dev libboost-chrono-dev libboost-atomic-dev libboost-date-time-dev libboost-math-dev libboost-serialization-dev libboost-fiber-dev libboost-context-dev``
+- *Debian/Ubuntu:* ``sudo apt-get install libboost-program-options-dev libboost-filesystem-dev libboost-system-dev libboost-thread-dev libboost-chrono-dev libboost-atomic-dev libboost-date-time-dev libboost-math-dev libboost-serialization-dev libboost-fiber-dev libboost-context-dev libboost-regex-dev``
 - *Arch Linux:* ``sudo pacman --sync boost``
 - *Spack:* ``spack install boost``
 - *from source:*

--- a/include/picongpu/CMakeLists.txt
+++ b/include/picongpu/CMakeLists.txt
@@ -165,10 +165,10 @@ endif()
 # Find Boost
 ################################################################################
 
-find_package(Boost 1.62.0 REQUIRED COMPONENTS program_options regex filesystem
+find_package(Boost 1.62.0 REQUIRED COMPONENTS program_options filesystem
                                               system math_tr1 serialization)
 if(TARGET Boost::program_options)
-    set(LIBS ${LIBS} Boost::boost Boost::program_options Boost::regex
+    set(LIBS ${LIBS} Boost::boost Boost::program_options
                      Boost::filesystem Boost::system Boost::math_tr1
                      Boost::serialization)
 else()

--- a/include/picongpu/initialization/ParserGridDistribution.cpp
+++ b/include/picongpu/initialization/ParserGridDistribution.cpp
@@ -1,0 +1,142 @@
+/* Copyright 2013-2019 Axel Huebl, Rene Widera, Benjamin Worpitz
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "picongpu/initialization/ParserGridDistribution.hpp"
+
+#include <pmacc/verify.hpp>
+#include <cstdint>
+#include <vector>   // std::vector
+#include <string>   // std::string
+#include <iterator> // std::distance
+
+#include <boost/regex.hpp>
+#include <boost/lexical_cast.hpp>
+
+
+namespace picongpu
+{
+
+    ParserGridDistribution::ParserGridDistribution( std::string const s )
+    {
+        parsedInput = parse( s );
+    }
+
+    uint32_t
+    ParserGridDistribution::getOffset( uint32_t const devicePos, uint32_t const maxCells ) const
+    {
+        value_type::const_iterator iter = parsedInput.begin();
+        // go to last device of these n subdomains extent{n}
+        uint32_t i = iter->count - 1u;
+        uint32_t sum = 0u;
+
+        while( i < devicePos )
+        {
+            // add last subdomain
+            sum += iter->extent * iter->count;
+
+            ++iter;
+            // go to last device of these n subdomains extent{n}
+            i += iter->count;
+        }
+
+        // add part of this subdomain that is before me
+        sum += iter->extent * ( devicePos + iter->count - i - 1u );
+
+        // check total number of cells
+        uint32_t sumTotal = 0u;
+        for( iter = parsedInput.begin(); iter != parsedInput.end(); ++iter )
+            sumTotal += iter->extent * iter->count;
+
+        PMACC_VERIFY( sumTotal == maxCells );
+
+        return sum;
+    }
+
+    uint32_t
+    ParserGridDistribution::getLocalSize( uint32_t const devicePos ) const
+    {
+        value_type::const_iterator iter = parsedInput.begin();
+        // go to last device of these n subdomains extent{n}
+        uint32_t i = iter->count - 1u;
+
+        while( i < devicePos )
+        {
+            ++iter;
+            // go to last device of these n subdomains extent{n}
+            i += iter->count;
+        }
+
+        return iter->extent;
+    }
+
+    void
+    ParserGridDistribution::verifyDevices( uint32_t const numDevices ) const
+    {
+        uint32_t numSubdomains = 0u;
+        for( SubdomainPair const & p : parsedInput )
+            numSubdomains += p.count;
+
+        PMACC_VERIFY( numSubdomains == numDevices );
+    }
+
+    ParserGridDistribution::value_type
+    ParserGridDistribution::parse( std::string const s ) const
+    {
+        boost::regex regFind( "[0-9]+(\\{[0-9]+\\})*",
+                              boost::regex_constants::perl );
+
+        boost::sregex_token_iterator iter( s.begin( ), s.end( ),
+                                           regFind, 0 );
+        boost::sregex_token_iterator end;
+
+        value_type newInput;
+        newInput.reserve( std::distance( iter, end ) );
+
+        for(; iter != end; ++iter )
+        {
+            std::string pM = *iter;
+
+            // find count n and extent b of b{n}
+            boost::regex regCount(
+                "(.*\\{)|(\\})",
+                boost::regex_constants::perl
+            );
+            std::string count = boost::regex_replace( pM, regCount, "" );
+
+            boost::regex regExtent(
+                "\\{.*\\}",
+                boost::regex_constants::perl
+            );
+            std::string extent = boost::regex_replace( pM, regExtent, "" );
+
+            // no count {n} given (implies one)
+            if( count == *iter )
+                count = "1";
+
+            const SubdomainPair g = {
+                boost::lexical_cast< uint32_t > ( extent ),
+                boost::lexical_cast< uint32_t > ( count )
+            };
+            newInput.emplace_back( g );
+        }
+
+        return newInput;
+    }
+
+} // namespace picongpu

--- a/include/picongpu/initialization/ParserGridDistribution.cpp
+++ b/include/picongpu/initialization/ParserGridDistribution.cpp
@@ -25,8 +25,7 @@
 #include <string>   // std::string
 #include <iterator> // std::distance
 
-#include <boost/regex.hpp>
-#include <boost/lexical_cast.hpp>
+#include <regex>
 
 
 namespace picongpu
@@ -98,12 +97,14 @@ namespace picongpu
     ParserGridDistribution::value_type
     ParserGridDistribution::parse( std::string const s ) const
     {
-        boost::regex regFind( "[0-9]+(\\{[0-9]+\\})*",
-                              boost::regex_constants::perl );
+        std::regex regFind(
+            R"([0-9]+(\{[0-9]+})*)",
+            std::regex::egrep
+        );
 
-        boost::sregex_token_iterator iter( s.begin( ), s.end( ),
+        std::sregex_token_iterator iter( s.begin( ), s.end( ),
                                            regFind, 0 );
-        boost::sregex_token_iterator end;
+        std::sregex_token_iterator end;
 
         value_type newInput;
         newInput.reserve( std::distance( iter, end ) );
@@ -113,25 +114,25 @@ namespace picongpu
             std::string pM = *iter;
 
             // find count n and extent b of b{n}
-            boost::regex regCount(
-                "(.*\\{)|(\\})",
-                boost::regex_constants::perl
+            std::regex regCount(
+                R"((.*\{)|(}))",
+                std::regex::egrep
             );
-            std::string count = boost::regex_replace( pM, regCount, "" );
+            std::string count = std::regex_replace( pM, regCount, "" );
 
-            boost::regex regExtent(
-                "\\{.*\\}",
-                boost::regex_constants::perl
+            std::regex regExtent(
+                R"(\{.*})",
+                std::regex::egrep
             );
-            std::string extent = boost::regex_replace( pM, regExtent, "" );
+            std::string extent = std::regex_replace( pM, regExtent, "" );
 
             // no count {n} given (implies one)
             if( count == *iter )
                 count = "1";
 
             const SubdomainPair g = {
-                boost::lexical_cast< uint32_t > ( extent ),
-                boost::lexical_cast< uint32_t > ( count )
+                static_cast< uint32_t > ( std::stoul(extent) ),
+                static_cast< uint32_t > ( std::stoul(count) )
             };
             newInput.emplace_back( g );
         }

--- a/include/picongpu/initialization/ParserGridDistribution.hpp
+++ b/include/picongpu/initialization/ParserGridDistribution.hpp
@@ -19,13 +19,9 @@
 
 #pragma once
 
-#include <pmacc/verify.hpp>
 #include <vector>   // std::vector
 #include <string>   // std::string
-#include <iterator> // std::distance
-
-#include <boost/regex.hpp>
-#include <boost/lexical_cast.hpp>
+#include <cstdint>
 
 
 namespace picongpu
@@ -52,41 +48,10 @@ private:
     using value_type = std::vector< SubdomainPair >;
 
 public:
-    ParserGridDistribution( std::string const s )
-    {
-        parsedInput = parse( s );
-    }
+    ParserGridDistribution( std::string const s );
 
     uint32_t
-    getOffset( uint32_t const devicePos, uint32_t const maxCells ) const
-    {
-        value_type::const_iterator iter = parsedInput.begin();
-        // go to last device of these n subdomains extent{n}
-        uint32_t i = iter->count - 1u;
-        uint32_t sum = 0u;
-
-        while( i < devicePos )
-        {
-            // add last subdomain
-            sum += iter->extent * iter->count;
-
-            ++iter;
-            // go to last device of these n subdomains extent{n}
-            i += iter->count;
-        }
-
-        // add part of this subdomain that is before me
-        sum += iter->extent * ( devicePos + iter->count - i - 1u );
-
-        // check total number of cells
-        uint32_t sumTotal = 0u;
-        for( iter = parsedInput.begin(); iter != parsedInput.end(); ++iter )
-            sumTotal += iter->extent * iter->count;
-
-        PMACC_VERIFY( sumTotal == maxCells );
-
-        return sum;
-    }
+    getOffset( uint32_t const devicePos, uint32_t const maxCells ) const;
 
     /** Get local Size of this dimension
      *
@@ -94,21 +59,7 @@ public:
      *  \return uint32_t with local number of cells
      */
     uint32_t
-    getLocalSize( uint32_t const devicePos ) const
-    {
-        value_type::const_iterator iter = parsedInput.begin();
-        // go to last device of these n subdomains extent{n}
-        uint32_t i = iter->count - 1u;
-
-        while( i < devicePos )
-        {
-            ++iter;
-            // go to last device of these n subdomains extent{n}
-            i += iter->count;
-        }
-
-        return iter->extent;
-    }
+    getLocalSize( uint32_t const devicePos ) const;
 
     /** Verify the number of subdomains matches the devices
      *
@@ -118,14 +69,7 @@ public:
      * \param[in] numDevices number of devices for this dimension
      */
     void
-    verifyDevices( uint32_t const numDevices ) const
-    {
-        uint32_t numSubdomains = 0u;
-        for( SubdomainPair const & p : parsedInput )
-            numSubdomains += p.count;
-
-        PMACC_VERIFY( numSubdomains == numDevices );
-    }
+    verifyDevices( uint32_t const numDevices ) const;
 
 private:
     value_type parsedInput;
@@ -140,48 +84,7 @@ private:
      * \return std::vector<SubdomainPair> with 2x uint32_t (extent, count)
      */
     value_type
-    parse( std::string const s ) const
-    {
-        boost::regex regFind( "[0-9]+(\\{[0-9]+\\})*",
-                              boost::regex_constants::perl );
-
-        boost::sregex_token_iterator iter( s.begin( ), s.end( ),
-                                           regFind, 0 );
-        boost::sregex_token_iterator end;
-
-        value_type newInput;
-        newInput.reserve( std::distance( iter, end ) );
-
-        for(; iter != end; ++iter )
-        {
-            std::string pM = *iter;
-
-            // find count n and extent b of b{n}
-            boost::regex regCount(
-                "(.*\\{)|(\\})",
-                boost::regex_constants::perl
-            );
-            std::string count = boost::regex_replace( pM, regCount, "" );
-
-            boost::regex regExtent(
-                "\\{.*\\}",
-                boost::regex_constants::perl
-            );
-            std::string extent = boost::regex_replace( pM, regExtent, "" );
-
-            // no count {n} given (implies one)
-            if( count == *iter )
-                count = "1";
-
-            const SubdomainPair g = {
-                boost::lexical_cast< uint32_t > ( extent ),
-                boost::lexical_cast< uint32_t > ( count )
-            };
-            newInput.emplace_back( g );
-        }
-
-        return newInput;
-    }
+    parse( std::string const s ) const;
 
 };
 

--- a/include/pmacc/debug/abortWithError.hpp
+++ b/include/pmacc/debug/abortWithError.hpp
@@ -28,6 +28,7 @@
 
 namespace pmacc
 {
+namespace{
     /** abort program with an exception
      *
      * This function always throws a `runtime_error`.
@@ -55,4 +56,5 @@ namespace pmacc
             msg
         );
     }
+}
 }

